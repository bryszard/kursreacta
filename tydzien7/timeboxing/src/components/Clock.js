import React from 'react';
import PropTypes from 'prop-types';

import { extractTimeFromMilliseconds } from '../lib/time.js';

const Clock = ({ remainingTime, text }) => {
    const limitRemainingTime = (remainingTime) => {
        const maxTime = (24 * 60 * 60 * 1000) - 1

        return Math.min(Math.max(remainingTime, 0), maxTime);
    }

    const remaining = limitRemainingTime(remainingTime)
    const { hours, minutes, seconds, milliseconds } = extractTimeFromMilliseconds(remaining);

    return (
        <h2>{ `${text}${hours}:${minutes}:${seconds}:${milliseconds}` }</h2>
    )
}

// function NonNegativeNumberType(props, propName, componentName) {
//   if (props[propName] < 0) {
//     return new Error(`Invalid group '${propName}' issued to component '${componentName}'. It has to be greater or equal to 0.`)
//   }
// }

Clock.defaultProps = {
  text: ''
}

Clock.propTypes = {
  remainingTime: PropTypes.number.isRequired,
  text: PropTypes.string
}

export default Clock;
