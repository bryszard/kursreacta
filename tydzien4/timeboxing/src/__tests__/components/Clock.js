import React from 'react';
import Clock from '../../components/Clock';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

describe('<Clock />', () => {
  it('sets text to empty string as default', () => {
    expect(<Clock remainingTime={10} />).toEqual(<Clock remainingTime={10} text="" />)
  })

  describe('when rendered component with ReactDOM', () => {
    let root;

    beforeEach(() => {
      root = document.createElement('div');
      ReactDOM.render(<Clock remainingTime={10000785} />, root)
    })

    it('renders h2 element', () => {
      expect(root.childNodes[0].nodeName).toEqual('H2')
    })

    it('renders time in HH:MM:SS:MMM format', () => {
      expect(root.childNodes[0].textContent).toEqual('02:46:40:785')
    })
  })

  describe('when rendered component with TestRenderer', () => {
    let clockRenderer;

    beforeEach(() => {
      clockRenderer = renderer.create(
        <Clock remainingTime={10000785} />
      )
    })

    it('renders properly', () => {
      expect(clockRenderer.toJSON()).toMatchSnapshot()
    })
  })
})
