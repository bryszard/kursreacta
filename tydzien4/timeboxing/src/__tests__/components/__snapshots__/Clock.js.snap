// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Clock /> when rendered component with TestRenderer renders properly 1`] = `
<h2>
  02:46:40:785
</h2>
`;
